docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -73,27 +74,57 @@ and includes all the core field members that are common across all transaction t
 Developers do not have to include `StdTxBase` if they wish, so it is meant to be
 used as an auxiliary type.
 
-### Signing
+### Signing & Transaction Submission
 
-Signing of a `Transaction` must be canonical across clients and binaries. In order
-to provide canonical representation of a `Transaction` to sign over, clients must
+To provide an acceptable user experience for client-side developers, we separate
+transaction encoding and signing. While every app defines its own transaction
+types for encoding purposes, we define a single standard signing `SignDoc` to be used
+across apps that leverages `google.protobuf.Any` to handle interface types. `Any`
+can be used to wrap any protobuf message by encoding both its type URL and value.
+We avoid using `Any` during the encoding phase because these URLs can be long
+and we want to keep encoded messages small. We are not, however, sending the
+`SignDoc` (encoded in JSON) over the wire or storing it on nodes,
+therefore using `Any` for signing does not add any encoding overhead.
+
+```proto
+message SignDoc {
+  StdSignDocBase base = 1;
+  repeated google.protobuf.Any msgs = 2;
+}
+```
+
+Signing of a `SignDoc` must be canonical across clients and binaries. In order
+to provide canonical representation of a `SignDoc` to sign over, clients must
 obey the following rules:
 
-- Encode `SignDoc` (see below) via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
+- Encode `SignDoc` via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
   - Default must be stripped from the output!
   - JSON keys adhere to their Proto-defined field names.
 - Generate canonical JSON to sign via the [JSON Canonical Form Spec](https://gibson042.github.io/canonicaljson-spec/).
   - This spec should be trivial to interpret and implement in any language.
 
-```Protobuf
-// app/codec/codec.proto
+Because signing and encoding are separated and apps will know how to encode
+a signing messages, we can provide a gRPC/REST endpoint that allows for generic
```

Comment:
```
we can provide a gRPC/REST endpoint that allows for generic
app-independent transaction submission. 
```

I don't understand what this means. Are you referring to a type to be used to generate a JSON blob which when provided to `/tx/broadcast` will be JSON decoded and then re-encoded using the app-specific tx encoding scheme?

Reply:
It's not json, it's protobuf via gRPC. Basically this AnyTransaction type which is not app dependent

Reply:
And yes, the rpc endpoint would re-encode this in the app level transaction that uses a `oneof`

Reply:
> It's not json, it's protobuf via gRPC

This makes sense. But in this case, isn't the term "REST" in "we can provide a gRPC/REST endpoint" misleading in this paragraph? I mean RPC and REST are entirely different API concepts and when creating a bridge from REST to gRPC, it would require more explanation.

Reply:
I still don't follow this @aaronc. I think it can be made more clear or explicit.

Reply:
I will remove the part about REST here and just say gRPC because REST refers to grpc-gateway and that's a separate topic.